### ISSUE TITLE
DuckDB - can it be replaced with other databases at the server?

### ISSUE THREAD

**Issue by KrishnaPG** [uwdata/mosaic#133]
This is a very useful package. The ReadMe says, "Explore massive datasets", but given that DuckDB is just a single file-based DB at the server (as a persistent data source), would it be able to handle large storage volumes? For example, continuously growing OLAP records every hour.

- Wondering how important DuckDB is for this whole thing to work. Is it tightly integrated with front-end? or can it be replaced with other large-scale databases?

For example, how easy / difficult it would be to make this work with, say Clickhouse at the backend (as the data source)?

**Comment by jheer** [comment]
At this point, we have only worked with and tested against DuckDB. We've also driven to larger scale datasets internally by running on a server machine with a large amount of RAM. Use of alternative backend DBs should work in theory, but would require additional engineering:

1. A new database connector would need to be written that connects to a database instance (e.g., Clickhouse or whatever), issues queries, and returns results (including in Arrow format). How easy or hard this is to author will depend on the specifics of the database.

2. Double check that the SQL generated by Mosaic is compatible with the target DB. Most components strive to follow common SQL standards, but we haven't tested this. Specific types/casts, references that span statements (e.g., referencing derived SELECT columns as GROUP BY criteria), and other syntactic bits may vary across engines. In any case, the file loading commands (for CSV, JSON, and Parquet) are definitely DuckDB-specific.

So in short: not yet. But I hope the above helps add some context!

**Comment by KrishnaPG** [comment]
Thanks, @jheer 

To understand the flow better,

- the front-end generates the query and sends it to the backend server. Is that correct?
  - that is, the backend server just have to forward the query to the DB and get the result (assuming the received SQL is in correct format)

If that is the case, is there any abstraction layer at the front-end to customize the generated SQL for various formats/backends? For example, generating GraphQL queries instead of plain SQL etc. 

That would make it easy for us to create adapters for various different backends.

The other option is, targeting something like [cube.dev](https://cube.dev/) as the standard backend (so that the various adapters can be taken advantage of).

**Comment by jheer** [comment]
Yes, that is largely correct. The Mosaic coordinator passes a query request to the current database connector (which is also client-side JS code). The connector is responsible for communication with the database. The existing connectors work with DuckDB-WASM i- browser (`wasm`), Mosaic's DuckDB server (from Mosaic's `duckdb` package) either over HTTP (`rest`) or Web Sockets (`socket`), or with Python DuckDB via Jupyter Comms. You can find implementations for the first three connectors within the `core` package. The Jupyter/Python connector is part of the `widget` package.

As for the abstraction layer, we currently assume that queries are either SQL strings or structured objects created with the Mosaic `sql` package utilities (which coerce to a SQL string). Note that data cube indexes will only be created with structured `Query` objects, as this permits us to analyze a query without having to parse it. Beyond that, the core architectural pieces could support alternative query languages, but all participating clients would have to generate those appropriately.

If you're interested in digging deeper, one route might be to look at the `sql` package's `Query` class and determine if you might translate the structured objects we already use into appropriate queries in your desired backing query language.

**Comment by KrishnaPG** [comment]
Thanks @jheer That is helpful.
